bestiary/core/api.py

```diff
@@ -0,0 +1,168 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2014-2020 Bitergia
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+# Authors:
+#     Santiago Dueñas <sduenas@bitergia.com>
+#     Miguel Ángel Fernández <mafesan@bitergia.com>
+#
+
+import django.db.transaction
+
+from .db import (find_ecosystem,
+                 add_ecosystem as add_ecosystem_db,
+                 update_ecosystem as update_ecosystem_db,
+                 delete_ecosystem as delete_ecosystem_db)
+from .errors import AlreadyExistsError, InvalidValueError, NotFoundError
+from .log import TransactionsLog
+from .utils import validate_name
+
+
+@django.db.transaction.atomic
+def add_ecosystem(ctx, name, title=None, description=None):
+    """Add an ecosystem to the registry.
+
+    This function adds an ecosystem to the registry.
+    It checks first whether the ecosystem is already on the registry.
+    When it is not found, the new ecosystem is added. Otherwise,
+    it raises a 'AlreadyExistsError' exception to notify that the ecosystem
+    already exists.
+
+    :param ctx: context from where this method is called
+    :param name: name of the ecosystem
+    :param title: title of the ecosystem
+    :param description: description of the ecosystem
+
+    :raises InvalidValueError: raised when name is None or an empty string
+    :raises TypeError: raised when name is not a string or a NoneType
+    :raises AlreadyExistsError: raised when the ecosystem already exists
+        in the registry
+    """
+    if name is None:
+        raise InvalidValueError(msg="'name' cannot be None")
+    if name == '':
+        raise InvalidValueError(msg="'name' cannot be an empty string")
+    if title == '':
+        raise InvalidValueError(msg="'title' cannot be an empty string")
+    if description == '':
+        raise InvalidValueError(msg="'description' cannot be an empty string")
+
+    trxl = TransactionsLog.open('add_ecosystem', ctx)
+
+    # Check if the name fulfills the requirements
+    try:
+        validate_name(name)
+    except ValueError as e:
+        raise InvalidValueError(msg=str(e))
```

Comment:
This should be tested in `db` because `validate_name` calls to `validate_field`. So it will be testing the same thing twice.